Document my ZST pointers asynchronous API design concept for libtock-rs.

jrvanwhy opened the pull request:
I have mentioned that I am experimenting with some ideas for lightweight asynchronous library design. Until now, I haven't described those ideas, because I was not confident that they are workable and *they are surprisingly difficult to describe*.

Here they are!

At this point, I am confident that we can construct a usable, lightweight syscall driver interface using my "ZST pointer" idea. Note that I haven't nailed down many of the details (such as which approach(s) we should take for handling `Sync`, or whether to use the generic traits I proposed), and those should be discussed in `libtock-rs` issues and PRs.

[Rendered](https://www.github.com/tock/design-explorations/tree/static-clients/zst_pointer_async)

Commit:
Document my ZST pointers asynchronous API design concept for libtock-rs.

jrvanwhy commented:
@torfmaster @Woyten

I intend to eventually incorporate some of these ideas into `libtock-rs`. There are a lot of details to work out, including but not limited to:

1. What crate structure to use (Add the drivers to `libtock-core`? Add the drivers to a new crate? Are the crates in a separate crate?)
1. Do we want to use the generic traits, and if so, what shape should they take?
1. Do we want to have callback marker types?

We should discuss those details in issues and design doc PRs in the `libtock-rs` repository.

torfmaster commented on zst_pointer_async/README.md:
```diff
@@ -0,0 +1,357 @@
+using Rust’s [Future](https://doc.rust-lang.org/core/future/trait.Future.html)
+trait to construct asynchronous APIs. Unfortunately, the analysis at [Futures
+versus no-Futures Size Comparison](../size_comparison) shows that the Future
+trait has size costs that are too high for many use cases of `libtock-rs`. As a
```
I would not agree to that extend. The point there is _a_ implementation of _a_ future based api which produces larger binaries at small scale. However, this implementation doesn't relate to the current future based api of `libtock-rs`.

torfmaster replied:
- [ ] Clarify this point

jrvanwhy replied:
What exactly are you looking for? I expect that the costs are too high for most of Google's use cases, so that alone probably counts as "many" (at least as a fraction of all `libtock-rs` use cases).

I don't see how futures can be implemented in a more lightweight mannen than they were in that API. If we don't store a `Waker` (`libtock-rs`'s current design), we lose the ability to use existing futures (e.g. `futures::stream::FuturesUnordered`).

torfmaster commented on zst_pointer_async/README.md:
```diff
@@ -0,0 +1,357 @@
+of its major concepts:
+
+1. **A static graph of interacting state machines:** The Tock kernel has a
+   statically-specified graph of objects that interact with each other. It is
```
Can you elobarate on this? I don't understand this. Can you define "static", "graph" and "objects"? Do you mean graphs in the mathematical sense? What do you mean by objects? How can state machines interact (they are data structures)? Can you give a reference to a place in the code where this graph is defined?

jrvanwhy replied:
I mean "graph" in the mathematical sense. "static" means the graph doesn't change at runtime. I'll add a clarification to "objects", but this is objects as in OOP.

State machines interact when state transitions affect each other. In code, this is implemented in the form of method calls/message passing (these are not distinct -- this whole construct looks a lot like Smalltalk OOP).

I'll put together a client example, which will show graph instantiation.

torfmaster commented on zst_pointer_async/README.md:
```diff
@@ -0,0 +1,357 @@
+
+1. **A static graph of interacting state machines:** The Tock kernel has a
+   statically-specified graph of objects that interact with each other. It is
+   uncommon to use dynamic handles (which is the norm when using futures), which
```
What do you mean by dynamic handles? Dynamic dispatch? Why do futures force you to use dynamic dispatch?

jrvanwhy replied:
Not necessarily dynamic dispatch, but in some form there need to be dynamic connections between the futures (which talk to multiple drivers) and kernel callbacks (which are static, and talk to multiple futures).

torfmaster commented on zst_pointer_async/README.md:
```diff
@@ -0,0 +1,357 @@
+// reference to the Foo is injected dynamically. Bar's client is injected fully
+// dynamically.
+struct Bar<F: Foo> {
+    foo: &F,  // 1 word
```
1 word when compiled in debug mode.

jrvanwhy replied:
This structure is meant to be used in a static context. Unless eliminated through dead code elimination, this will never be optimized away.

Every single statement of the form "writing X will cost Y bytes" has an implicit "unless it is entirely optimized away". I think the statement is fine as-is.

torfmaster commented on zst_pointer_async/README.md:
```diff
@@ -0,0 +1,357 @@
+   operation, which returns without blocking, then `B` calls `A` back when the
+   operation is finished.
+
+For example, in the Tock kernel, a connection between an object `Bar` and an
```
A little bit more of explanation (and maybe more descriptive names than foo, bar, baz) would be helpful here.
Maybe like this:
 * during operation "bar" Bar wants to request "foo" from Foo by calling foo()
 * However, foo is asynchronous so a callback is installed which notifies bar when foo is done with its work
 * the mechanism which establishes this connection is the BarClient
 * ...

jrvanwhy replied:
I added a description of how the connection is set up in the board's reset handler.

torfmaster commented on zst_pointer_async/README.md:
```diff
@@ -0,0 +1,357 @@
+    client: &dyn BarClient,  // 2 words
+}
+
+impl<F: Foo> FooClient for Bar<F> { ... }
```
Actuallly, this implementation would be quite descriptive to understand what is going on.

- [ ] Consider expanding the impl block.

jrvanwhy replied:
I expanded it as much as I could and added a comment. Is that good enough?

torfmaster commented on zst_pointer_async/README.md:
```diff
@@ -0,0 +1,357 @@
+1. A vtable in flash with 3 words Tock doesn't use (size, alignment, and
+   destructor).
+1. Prevents those calls from being inlined. LLVM has devirtualization to
+   alleviate this but it is not very effective in practice (in part because it
```
Do you have references for that assertion?

- [x] Give references for the assertion.

jrvanwhy replied:
Not really, it's mostly a practical observation I've made while dealing with the `core::fmt` code bloat. There is https://github.com/rust-lang/rust/issues/45774, I'll add a link to that.

jrvanwhy replied:
> What I currently don't understand: How does this relate to `libtock-rs` (in contrast to the kernel)? I find code similarly to the capsule code, however at least superficially `libtock-core` does something completely different for subscriptions.

I intend to add non-futures-based abstractions over Tock's system call drivers to `libtock-rs`. I'm trying to make those abstractions as lightweight as possible while achieving a reasonable level of usability. This document describes the fundamental concepts that I intend to use in the libtock-rs design, as well as some of the design constraints I have found that will constrain `libtock-rs`'s implementation.

`libtock-core` doesn't currently have any system call drivers, so this has no relationship to the current implementation.

> What does this mean in terms of memory usage? What is the advantage of your approach over the current approach?

Memory usage: Except for virtualization (which is no heavier than the kernel's virtualization implementation), none of the constructs mentioned in this document have any impact or memory or code size.

torfmaster commented on zst_pointer_async/README.md:
````diff
@@ -0,0 +1,357 @@
+```
+// FooPtr is implemented by a type that knows how to find an implementation of
+// `Foo`.
+trait FooPtr: Copy {
````
What does Ptr mean? Pointer? 

- [ ] Please choose more descriptive names here.

torfmaster replied:
Why do you need a `Copy` implementation here? 

- [x] Document it or or omit it.

jrvanwhy replied:
I documented the meaning of `Ptr` and the purpose of `Copy`.

torfmaster commented on zst_pointer_async/lw/time.rs:
```diff
@@ -0,0 +1,123 @@
+// wraparound, but the numbers returned by the kernel would match the second
+// sequence. Therefore, we need all queued alarm callbacks to run less than
+// 2^32 - UPDATE_PERIOD ticks after the alarm fires. We don't have a way to make
+// that happen (as Tock is not a RTOS) so we simply hope such a large delay
```
Typo: an RTOS

jrvanwhy replied:
Fixed in the latest commit.

torfmaster commented:
@jrvanwhy I'm currently trying to read through this but this is a _lot_. Can you at least provide one example how to use this api?

torfmaster commented:
What I currently don't understand: How does this relate to `libtock-rs` (in contrast to the kernel)? I find code similarly to the capsule code, however at least superficially `libtock-core` does something completely different for subscriptions. What does this mean in terms of memory usage? What is the advantage of your approach over the current approach?

jrvanwhy commented:
> @jrvanwhy I'm currently trying to read through this but this is a _lot_. Can you at least provide one example how to use this api?

I'm putting together an example. I intend to push it (and the other changes I said I'd make) tomorrow.

Commit:
Add an example client app, rename `Forwarder` to `AsyncClientPtr` to match the README's naming scheme, and apply a few of @torfmaster's suggestions.

jrvanwhy commented:
> @jrvanwhy I'm currently trying to read through this but this is a _lot_. Can you at least provide one example how to use this api?

There is now an example of client code at `zst_pointer_sync/client_example.rs`.

Commit:
More documentation improvements:

1. Better describe how the kernel is structured.
2. Clarify the design of FooPtr in the first ZST pointer design.
